### DISCUSSION TITLE
Difference between "Pass through" and "Nothing" action

### DISCUSSION THREAD

**Discussion by shuvashish76** [koreader/koreader#12702] [Q&A]
Every gesture has these 2 actions on top. But I don't see any difference between the two. What is the meaning and use of "Pass through" gestutre-action?

**Comment by benoit-pierre** [comment]
* "Nothing" → swallow the action
* "Passthrough" → another handler might act on it

**Reply by shuvashish76** [reply]
So passthrough is useful only if you use multiple actions for one gesture?
Any use-case example?

**Reply by benoit-pierre** [reply]
"Pass through" is useful to keep the default behavior, like text selection, for example when long pressing near the corners.

"Nothing" can be useful to avoid accidental actions, "pass through" to reduce conflicts.

For example, on my Kindle: changing pages on taps only, reserving swipes for following links and to go back to the previous location. This way I can tap to change pages without issues even on pages with a lot of links (like a PDF with a lot of references), and only follow links when I intend to. Long press near corners set to pass through so text selection work, but some specific actions on corner taps.

**Reply by jonnyl2** [reply]
Or another example: if 'tap corner' is set to pass-through, a tap in the corner will still turn the page (provided taps are enabled for page turns, of course). If it's set to 'Nothing', nothing happens and taps to turn the page must be located outside that corner touch area. 

I was initially also somewhat confused by those two options. Instead of just "Nothing", "Do nothing" might be slightly clearer wording. And instead of "Pass through", "Ignore" might be better.

**Reply by benoit-pierre** [reply]
Or "Suppress" and "Ignore".

**Reply by mergen3107** [reply]
That distinction is great! Once I got my first eink KOReader device with flush screen, I found out how sloppy my palm is when I hold a device in hand.

Whenever I slightly moved, I accidentally tapped a corner and it advanced a page. Now I have them as "Nothing" - no more unwanted page turns :D

**Reply by ichnilatis-gr** [reply]
> And instead of "Pass through", "Ignore" might be better.

This would be desirable as this expression is not translated very well in some languages (e.g. Greek) to convey this meaning.

**Reply by mergen3107** [reply]
I would disagree. 

"Ignore" implies that, say, a corner tap will be ignored, whereas in fact it is not ignored, but used as a usual tap for page turns. 

"Do not modify" might be better maybe

**Reply by Frenzie** [reply]
You're in front of the gate. If you're ignored, you can't do anything in the city. If they let you (pass) through, you're in the city. Ignore is a synonym for nothing, the exact opposite of pass through.

Of course here it was intended as ignoring you while you enter the gate, but that's ambiguous. Pass through is significantly clearer.

If the Greek equivalent for "let him through" is "ignore him" there's nothing wrong with translating it that way. But the English is fine as it is.

**Reply by ichnilatis-gr** [reply]
My comment didn't mean that I prefer the term "Ignore", it rather expressed a desire for a new term, because the term "pass through" is not very clear.

**Reply by Frenzie** [reply]
I'm afraid I can't agree with that. Perhaps it's "nothing" that's unclear?

**Reply by jonnyl2** [reply]
"Pass through" is clear when explaining what the gesture does from one programmer to another. But a regular end user doesn't think in terms of "what chain of events will be effected when I perform this gesture," but rather, "what is assigned to this gesture," or simply "what happens." In this context, "pass through" just elicits a _"...what's a pass through?"_ response.

"Nothing" is ambiguous because it could mean one of two things: 1) Nothing assigned, or 2) No action / Nothing happens. The first one would be closer to the current "Pass through," for which I suggested "Ignore" because that's effectively what it is from the user's perspective: KOReader just ignores this particular gesture setting (not the gesture itself, which again, could be ambiguous I admit).

So, another attempt:

Pass through -> Nothing assigned (pass through gesture)
Nothing -> Do nothing (consume gesture)

Just FTR, I don't mind if it's just left the way it is. But I find this discussion interesting and I can appreciate that some users are confused by the naming, because I briefly went through the same experience as a new user. 

Perhaps @offset-torque has some opinion on this as well. I'd be interested in his take.

**Reply by ichnilatis-gr** [reply]
I had once ask the meaning of the term "Pass through" in [MobileRead forum](https://www.mobileread.com/forums/showthread.php?t=333285) and I admit that even now I haven't fully understood its meaning. I'm sorry about that. I know it's my problem...

So if I have assigned an action to a gesture in the Reader, do  I have to set this gesture to "Pass through" in the File Browser, if I don't want an action to be assigned to that gesture?

**Reply by jonnyl2** [reply]
No. The gesture settings for Reader and File browser are completely separate. 

Just think of the tap-areas on the screen as different layers; e.g. in the Reader: the right bottom corner is a layer on top of the layer for page turn taps. So if you tap in the bottom right corner and the gesture is set to "pass through," the gesture is passed through to the page turn layer, and the page still turns. Wheareas if you set it to "Nothing" then literally nothing happens and the gesture is consumed by the system and not further passed along.

**Reply by Frenzie** [reply]
> "Pass through" is clear when explaining what the gesture does from one programmer to another. 

No, a programmer says an event propagates. This wording is carefully considered plain English. If you ring the doorbell, when someone ignores you you're left outside. If someone lets you pass through, you're inside.

**Reply by jonnyl2** [reply]
Fair enough. I just meant conceptually.

**Reply by hius07** [reply]
"Do nothing" looks consistent.
https://github.com/koreader/koreader/blob/47ab7685aa683a538bc7906d9a2f89fb6e3e4080/frontend/ui/elements/common_settings_menu_table.lua#L682
https://github.com/koreader/koreader/blob/47ab7685aa683a538bc7906d9a2f89fb6e3e4080/frontend/apps/reader/modules/readerhighlight.lua#L366-L368

**Reply by Frenzie** [reply]
True, though that might be an argument against the "nothing" phrasing since it doesn't mean "drop"/"swallow".

**Reply by offset-torque** [reply]
This is a **"mental model of the system"** problem. I will explain but firstly, I totally agree with what **jonnyl2** said above. I agree all of his long reply so I am not quoting something specific. I will just expand on the same direction.

So according to the explanations in this thread, I made this diagram. It shows two different perspectives for a gesture, let's say **DOUBLE TAP ON THE TOP RIGHT CORNER**:
- Purple rectangle: How a **DEV** sees the system
- Green rectangle: How a **USER** sees the system

![koreader_passthrough_discussion](https://github.com/user-attachments/assets/b0db1f14-15c6-4d29-af5b-1031942775f8)


Concept of **"Pass through"** makes sense if you have the mental model of a dev who knows how KOReader works.

But from the perspective of a normal user, there are no different layers or sub-systems passing things to each other. They only see the KOReader application. They can guess what "Pass through" means but they can't know the possible target(s) of this passing through so it is understandably confusing.

**SOLUTION (IMHO)**
As a user I only care about one thing:
**What happens if I double tap the top right corner?**

And there are only 2 options:
- Something will happen, because there is an assigned action.
- Nothing will happen, because KOReader will ignore my taps.

**There is no third option from my perspective. Double tap either does something or nothing happens.**

As a user I don't care about the underlying mechanism. I don't want to know the internals because I don't need to know the internals. Also I can't know because it is not exposed or explained in the UI.

So don't show me a fuzzy **PASS THROUGH**. Show me the final action after the passing through because that is all I care.

As far as I know, there are no other user interfaces that expose input internals to the user like this. Everything from the game keybindings to the mouse button remapping works like this: 
- DEFAULT or USER DEFINED ACTION
- IGNORED/UNASSIGNED

Please correct me if I am wrong.

**BONUS**
If we remove the problematic PASS THROUGH, naming of IGNORE/DO NOTHING problem will be automatically solved too. Because these two options will merge into IGNORE which will mean **"don't react to my input when I do this tap/swipe/whatever"** as the users expected.

**Reply by hius07** [reply]
> If we remove the problematic PASS THROUGH

We need it. For example, tap on the upper right corner. If it is set to "Pass through", it would open the Upper menu, or Turn page, or do nothing, depending on the other settings in the Taps & gestures menu.

**Reply by offset-torque** [reply]
And that is the source of all this problem and confusion. There are more than one mechanism for setting the actions, gesture's page is not self contained if it is the correct term. I am not criticizing anyone, to clarify. Probably this system evolved out of necessity and built on some legacy code with workarounds. But the result is confusing from the viewpoint of users.

**Reply by Frenzie** [reply]
> So don't show me a fuzzy PASS THROUGH. Show me the final action after the passing through because that is all I care.

The final action is unknown. The final action is passed through to whatever else would be doing the action if it weren't intercepted by the corner. This is how hot corners work on all operating systems. What hot corners may not have is "do nothing". If anything should be removed it's "do nothing," but some people like that.

A possible alternative is `—` (as in unset).

**Reply by Frenzie** [reply]
> Concept of "Pass through" makes sense if you have the mental model of a dev who knows how KOReader works.

I object to this. You're explicitly setting corners to do something else other than what normally happens under the corner. You can't install a door where there used to be none and be surprised when the door stops you.

**Reply by Frenzie** [reply]
For example:
<img src=https://github.com/user-attachments/assets/a4cae511-602b-471d-a4b5-5bc1e4cea5d5 width=40%> <img src=https://github.com/user-attachments/assets/c38e3ca7-c1dc-42f8-a031-1882efc09504 width=40%>

**Reply by mergen3107** [reply]
> You're explicitly setting corners to do something else other than what normally happens under the corner.

So maybe the whole perspective of gestures needs to be slightly reworded?

Settings should say "Override gestures". 

Then you go in, select a gesture and see two checkmark options at the top:
- No (this will keep it as "Pass through")
- Yes
This selection will open another menu with specific actions (or maybe make gray out all other options when "No" is selected, and make them active when "Yes" is selected). And "Nothing" is just another option under this Yes, but maybe renamed ("Ignore/Consume this gesture")

**Reply by Frenzie** [reply]
I don't like extra actions. ;-)

But perhaps `override` isn't bad. For example:
<img src=https://github.com/user-attachments/assets/715a3a5c-34b9-4cf7-8c4b-c2e03f459f51 width=45%> <img src=https://github.com/user-attachments/assets/6d10f984-a0f8-4bbd-9d70-bad1f012f436 width=45%>

Edit: should be `drop gesture`

**Reply by offset-torque** [reply]
By the way I will add this information to the guide for curious users, also to be able to direct them if another question like this comes up, instead of explaining everything again. I tried to write my replies from the viewpoint of a normal user. Personally this issue doesn't bother me at all because I use only two gestures - one for the frontlight and one for my quick menu so I don't interact with the gesture manager much.

**Comment by offset-torque** [comment]
> The final action is unknown.

I can't remember any other user interface that works like this. Maybe I am not thinking very hard.

> This is how hot corners work on all operating systems.

This is from my current KDE Linux settings. Hot corner action is set from this popup. I don't have to go to any other page to set this corner. And nothing overrides this setting. Whatever I see and set in this single menu is what happens when I touch that corner. There is no confusion. There is no ambiguity, final action is always known.

![kde_screen_edges](https://github.com/user-attachments/assets/08d2edf4-7ca2-4f49-8135-1fd01b2af173)



> I object to this. You're explicitly setting corners to do something else other than what normally happens under the corner

Exactly what is happening in the picture above. If KDE decided to ship the default left corner action as "Peek at desktop" it would be still in this menu (and selected).

**Reply by offset-torque** [reply]
Sorry I split the discussion by accident. I reflexively wrote in the bottom box :)

**Comment by Frenzie** [comment]
Closing. The current situation is superior.
